design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,690 @@
+  represent.
+- _`Package` controller_: a controller responsible for observing events on
+  `Package` instances, creating new `PackageRevision` instances, and managing
+  which revision is active.
+- _`PackageRevision` controller_: a controller responsible for observing events
+  on `PackageRevision` instances, installing new CRDs /
+  `InfrastructureDefinition` / `InfrastructurePublication` / `Composition`
+  objects, starting package controllers, and cleaning up these objects when a
+  package is upgraded.
+
+These types and controllers overlap in functionality with some of the existing
+types and controllers that are currently present in the package manager
+ecosystem. However, their implementation should not immediately break any of the
+current features supported by the package manager. Removal of
+`ClusterPackageInstall`, `PackageInstall`, and `StackDefinition`, as well as
+refactor of `Package` and all controllers should follow a deprecation schedule
+that allows users to move to the new format.
+
+## Implementation
+
+This following sections describe the changes to the current package manager that
+are required to support the goals listed above.
+
+- [Current Workflow](#current-workflow)
+- [Future Workflow](#future-workflow)
+- [Phase 1](#phase-1)
+- [Phase 2](#phase-2)
+- [Phase 3](#phase-3)
+
+### Current Workflow
+
+In order to accurately describe the steps required for implementing this
+refactor. It is important to understand, at least at a high-level, the current
+workflow the package manager uses for installing a `Package`. We will use a
+`ClusterPackageInstall` for the example here as the namespace-scoped
+`PackageInstall` is intended to be removed as part of this refactor.
+
+1. User creates a `ClusterPackageInstall`.
+2. The _`ClusterPackageInstall` controller_ observes the `ClusterPackageInstall`
+   creation and creates a `Job` that runs a `Pod` with an initContainer running
+   the image supplied on the `ClusterPackageInstall` and a container running the
+   same image the package manager itself is running. The `initContainer` serves
+   to only copy the contents of the package directory from the supplied image
+   into a `Volume`. That `Volume` is then mounted on the `container` running the
+   package manager image, which is executed with the `unpack`
+   [argument](https://github.com/crossplane/crossplane/blob/a0d139f7cf269599ba916ed15af3fd68ffeabbdf/cmd/crossplane/package/unpack/unpack.go#L52).
+3. The `unpack` container walks the filepath of the package directory, using the
+   information provided to construct a `Package` object. It then
+   [writes](https://github.com/crossplane/crossplane/blob/a0d139f7cf269599ba916ed15af3fd68ffeabbdf/pkg/packages/unpack.go#L204)
+   the `Package` object and all CRDs in the package to
+   [stdout](https://github.com/crossplane/crossplane/blob/a0d139f7cf269599ba916ed15af3fd68ffeabbdf/cmd/crossplane/package/unpack/unpack.go#L53).
```

Comment:
I do like that format better. I think that one or maybe an iteration on that would be fine. This is also something that we could modify over time since it is not user-facing.